### ISSUE TITLE
[FEAT] Change the variable and function whose name is 'check' to other names. Or is there any way to avoid conflicts when 'check' is defined in other libraries when integrating with other libraries.

### ISSUE THREAD

**Issue by HakimHua** [pybind/pybind11#2591]
The name ‘check’ is used in many libraries, and even some libraries define the name as a global function (such as, in Unreal Engine 4, the 'check' function is defined as a macro), which causes conflicts and errors when integrating with pybind11.

I tried it : Changing the names of all the functions or variables named 'check' in pybind11 can avoid these problems. Is there any other more convenient mechanism to avoid this?

**Comment by bstaletic** [comment]
You'll have to provide more details. As far as I can tell, nothing named `check` is in the public API.

**Comment by YannickJadoul** [comment]
@bstaletic There's a (deprecated) `check` and `check_` in the `py::object` protocol.

@HakimHua changing this would probably break a lot of custom `py::object`s, though. I'd say that ours is nicely scoped and namespaced, and that all those libraries defining a common macro like `check` are completely wrong: 1) macros should be `ALL_CAPS`, 2) macros should most likely be namespaced as `UE4_CHECK` or so, if they're in public headers! We've also changed `NAMESPACE_BEGIN` into `PYBIND11_NAMESPACE_BEGIN` to fix this in pybind11.

**Comment by YannickJadoul** [comment]
> Or is there any way to avoid conflicts when 'check' is defined in other libraries when integrating with other libraries.

Yes, `#undef check` after including headers that pollute these names. Nórmally, global functions (not dirty macros, but functions) shouldn't be a problem? Or are they?

**Comment by bstaletic** [comment]
> There's a (deprecated) check and check_ in the py::object protocol.

Let me rephrase that. Any `check` that pybind uses is either a member function, a comment or a function-local variable. Pybind11 is not wrong here. We're not the standard library and we shouldn't defend against evil macros. Else all our identifiers would have to be `__ugly`.

**Comment by henryiii** [comment]
Unreal 4 is known to be horrible for compatibility. Here's the suggestion one poster has:

https://answers.unrealengine.com/questions/391017/constant-library-conflicts.html

```cpp
#pragma push_macro("CONSTEXPR")
#undef CONSTEXPR
#pragma push_macro("dynamic_cast")
#undef dynamic_cast
#pragma push_macro("check")
#undef check
#pragma push_macro("PI")
#undef PI
#include <that-3rd-party-library.hpp>
#pragma pop_macro("PI")
#pragma pop_macro("check")
#pragma pop_macro("dynamic_cast")
#pragma pop_macro("CONSTEXPR")
```

Yes, they redefine `dynamic_cast`.

The official answer is Unreal 4 does not support third party libraries, by the way.

> As we don't support those extra libraries, we most likely won't be making any direct changes to the source code to change this at this time. However, I would suggest looking at AllowWindowsPlatformTypes.h and HideWindowsPlatformTypes.h and how they are handling the macros defined there. This should be able to give you an idea of how to incorporate the Boost library's macros that are conflicting.

**Comment by YannickJadoul** [comment]
> Yes, they redefine `dynamic_cast`.

They do WHAT!?

**Comment by YannickJadoul** [comment]
Anyhow, on that note, I think we can close this issue? Unless there's problem with a global name in pybind11 or something like that (i.e., we did change `class module` into `class module_`, to avoid accidental collisions with C++20), concerning sane libraries that do not explicitly say they don't support other libraries, I don't think we can reasonably be expected to change anything?

Do reopen if we're overlooking something, please, @HakimHua!

**Comment by bstaletic** [comment]
Curiously enough, `#define`-ing existing keywords is UB.
